Add graceful server shutdown

andreiltd opened the pull request:
This patch adds graceful shutdown functionality to the Wasmtime shim server loop. The shutdown process works as follows:

- The code that spawns the top-level loop creates a `CancellationToken`, which is shared across all cancellable tasks.
- A tracker is used to manage any tasks spawned within the server loop.
- A signal handler is installed to listen for signals. Upon receiving the first `Interrupt` signal, it calls cancel and waits for the server task to finish gracefully, allowing all spawned tasks to complete.
- On receiving a second `Interrupt` signal, the server loop exits immediately.
- Similarly, upon receiving a `Terminate` signal, the server loop finishes abruptly.
- The return code indicates which of these shutdown scenarios occurred. If the server has finished abruptly the `128 + SIG` will be used. See https://tldp.org/LDP/abs/html/exitcodes.html for more details.

Mossaka commented on crates/containerd-shim-wasmtime/src/instance.rs:
```diff
@@ -302,9 +306,32 @@ where
-            tokio::select! {
-                status = self.execute_component_async(ctx, component, func, stdio) => { status }
-                sig = wait_for_signal() => { sig }
+            let mut done = false;
```
question: what is this used for?

andreiltd replied:
I think we can come up with a better name for this, like `force_shutdown`. This variable tracks if we already initiated a shutdown and upon receiving a second `ctrl_c` signal, we cancel waiting for the tasks and exit immediately.

Mossaka replied:
agreed that `force_shutdown` would be a better name

Mossaka commented on crates/containerd-shim-wasmtime/Cargo.toml:
```diff
@@ -6,10 +6,11 @@ edition.workspace = true
 hyper = { workspace = true }
 tokio = { workspace = true, features = ["signal", "macros"] }
-libc = { workspace = true }
+tokio-util = { workspace = true, features = ["rt"] }
```
I was reading the difference between `JoinSet` and `TaskTracker` but I didn't understand if our use cases require some of the unique properties that `TaskTracker` brings. If not, I'd suggest we go with `JoinSet` since it meant one less dependency for runwasi.

andreiltd replied:
The property we leverage is

> 4. A TaskTracker can be cloned to share it with many tasks.

The tracker is cloned and passed to handler to track the tasks spawned there as well:
https://github.com/containerd/runwasi/pull/705/files#diff-b2d23b367806b354ef48f11bbfd796edae1c54af02e2a28f83fbbdf273b955b6R70

I think `JoinSet` is not `Clone` but I will double check.

Mossaka commented:
Please consider adding integration tests that simulate different shutdown scenarios

jprendes commented on crates/containerd-shim-wasmtime/src/http_proxy.rs:
```diff
@@ -59,45 +62,65 @@ pub(crate) async fn serve_conn(
             }
-        });
+            res = listener.accept() => {
+                let (stream, _) = res?;
```
**question:** What does it mean for `res` to  be `Err`?
The current behaviour triggers a forced shutdown, is that what we want?
Otherwise we can do something like
```suggestion
                let Ok(stream, _) = res else {
                    log::info!("Error accepting new connection, triggering graceful shutdown.");
                    break;
                };
```

and follow the graceful shutdown path.

andreiltd replied:
Good catch! I now changed this and followed `axum` strategy to deal with errors in accept connection.

jprendes commented on crates/containerd-shim-wasmtime/src/instance.rs:
```diff
@@ -302,9 +306,32 @@ where
+
+            tokio::pin!(exec);
+
+            loop {
```
**comment:** I got a bit confused by this being a loop, but then realized this will only run one, and at most two iterations. I wonder if this can be rewritten without a loop.

andreiltd replied:
I think the problem is that we want to continue with server future after we receive a first signal. I 'm not sure if it's possible without loop.

Mossaka replied:
Can we do this to make sure that it at most loops twice? 
```rust
for _ in 0..2 {
  ...
}
```

andreiltd replied:
It is now rewritten without a loop. Thanks @jprendes for suggestion.

jprendes reviewed (commented):
Thanks for working on this @andreiltd !

andreiltd commented:
The last commit introduces a few changes:
- Added an integration test to verify forced shutdown using a terminate signal.
- Addressed errors in accepting connections by implementing a backoff strategy, which is a common pattern found in libraries such as `axum` and `mini-redis`.

Mossaka reviewed (dismissed):
Nice work, thanks for adding the integration test!

jprendes reviewed (dismissed):
LGTM

jprendes reviewed (approved):
Thanks @andreiltd !
